### PR TITLE
adding option for imagePullPolicy: Always to be optional

### PR DESCRIPTION
**Pull request by TonyRice**

### CODE REVIEW

File: asyncy/Apps.py [hunk 1]
```diff
@@ -284,14 +285,21 @@ def get(cls, app_id: str):
             """
             curs.execute(query, (app_id,))
             release = curs.fetchone()
+            if release is None:
+                glogger.warn(f'Cowardly refusing to deploy app '
+                     f'{app_id}. No release data found.')
```
Comment (wilzbach) [inline asyncy/Apps.py]: Should be an Error.

File: asyncy/Containers.py [hunk 2]
```diff
@@ -145,6 +145,7 @@ def is_service_reusable(cls, app, line):
         story.logger.info(f'Starting container {service}')
         container_name = cls.get_container_name(story.app, story.name,
                                                 line, service)
+
```
Comment (wilzbach) [inline asyncy/Containers.py]: Unrelated

File: asyncy/Kubernetes.py [hunk 3]
```diff
@@ -461,6 +461,9 @@ def format_ports(cls, ports: {int}):
 
         b16_service_name = base64.b16encode(service_name.encode()).decode()
 
+        if app.always_pull_images:
+            app.logger.debug(f'imagePullPolicy set to Always')
```
Comment (wilzbach) [inline asyncy/Kubernetes.py]: Escape values.

File: asyncy/Kubernetes.py [hunk 4]
```diff
@@ -498,7 +501,7 @@ def format_ports(cls, ports: {int}):
                                     }
                                 },
                                 'command': start_command,
-                                'imagePullPolicy': 'Always',
+                                'imagePullPolicy': 'Always' if app.always_pull_images else 'IfNotPresent',
```
Comment (wilzbach) [inline asyncy/Kubernetes.py]: Create a separate variable for this and use it in the logging above.

File: asyncy/Kubernetes.py [hunk 5]
```diff
@@ -135,7 +135,7 @@ def new_ssl_context(cls):
 
         cert = config.CLUSTER_CERT
         cert = cert.replace('\\n', '\n')
-        context.load_verify_locations(cadata=cert)
+        #context.load_verify_locations(cadata=cert)
```
Comment (judepereira) [inline asyncy/Kubernetes.py]: Why is this commented?

Comment (TonyRice) [reply asyncy/Kubernetes.py]: Was done by mistake, just reverted that. For some reason, ssllib errors out every time on my certificate data.

**Comment by judepereira** [comment]
@TonyRice Can you check why the tests fail?

**Comment by TonyRice** [comment]
> @TonyRice Can you check why the tests fail?

Missed a flag in the test, and I believe a few other unrelated test are failing. Fixing now.

**Comment by TonyRice** [comment]
> Looks like some formatting tests are failing. Could you fix them please?

On it!

**Comment by TonyRice** [comment]
Fixing conflict so this can be merged

**Comment by TonyRice** [comment]
@judepereira can you review the latest changes on this?

**Comment by judepereira** [comment]
Yes, will review today. 


Cheers,
Jude Pereira

File: asyncy/Kubernetes.py [hunk 6]
```diff
@@ -560,6 +561,11 @@ def get_liveness_probe(cls, app, service: str):
 
         b16_service_name = base64.b16encode(service_name.encode()).decode()
 
+        image_pull_policy = 'Always' if app.always_pull_images \
```
Comment (judepereira) [inline asyncy/Kubernetes.py]: Please use an if/else block here, instead of the inline if. It's far easier to look at.

File: asyncy/db/Database.py [hunk 7]
```diff
@@ -86,16 +87,17 @@ def get_release_for_deployment(cls, config, app_id):
...
-                environment=data['environment'],
-                stories=data['stories'],
-                maintenance=data['maintenance'],
-                app_dns=data['app_dns'],
-                state=data['state'],
-                deleted=data['deleted'],
-                owner_uuid=data['owner_uuid'],
-                owner_email=data['owner_email']
-            )
+        return Release(
```
Comment (judepereira) [inline asyncy/db/Database.py]: Looks like the indentation changed here. Is this correct?

File: tests/unit/App.py [hunk 8]
```diff
@@ -369,7 +373,7 @@ def test_app_init(magic, config, logger, env):
     app.config.ASYNCY_SYNAPSE_PORT = 9000
 
     expected_url = f'http://{app.config.ASYNCY_SYNAPSE_HOST}:' \
-                   f'{app.config.ASYNCY_SYNAPSE_PORT}/clear_all'
+        f'{app.config.ASYNCY_SYNAPSE_PORT}/clear_all'
```
Comment (judepereira) [inline tests/unit/App.py]: Are you running auto format on the entire file? If yes, please don't do so, and only run auto format on your changed code. IntelliJ does have this feature.

File: asyncy/Apps.py [hunk 9]
```diff
@@ -276,8 +276,9 @@ def get(cls, app_id: str):
                     config, app_id, release.app_name,
                     release.app_dns, release.version,
```
Comment (judepereira) [inline asyncy/Apps.py]: Can you rather refactor deploy_release to accept the release object instead? So then it'll just have three params, config, app_id, release.

**Comment by TonyRice** [comment]
@judepereira everything has been corrected and tested with local deployments. accidental re-formatting has been undone too.

**Comment by judepereira** [comment]
@TonyRice correct me if I'm wrong, but I don't see a test where the Kubernetes manifest is tested for Always vs IfNotPresent.

**Comment by TonyRice** [comment]
> @TonyRice correct me if I'm wrong, but I don't see a test where the Kubernetes manifest is tested for Always vs IfNotPresent.

You're not wrong, I'll add shortly

**Comment by TonyRice** [comment]
> Tests?

For some reason, I cannot resolve the conversation, @wilzbach can you approve this?

File: asyncy/Apps.py [hunk 10]
```diff
@@ -110,7 +117,8 @@ def get_app_config(cls, raw):
                     logger=logger,
                     stories=stories,
                     services=services,
-                    environment=environment,
+                    always_pull_images=release.always_pull_images,
+                    environment=release.environment,
```
Comment (wilzbach) [inline asyncy/Apps.py]: While you're doing this refactoring: why not pass `release` to `AppData`?

File: asyncy/Apps.py [hunk 11]
```diff
@@ -47,10 +48,16 @@ def get_app_config(cls, raw):
...
+    async def deploy_release(cls, config: Config, release: Release):
+        app_id = release.app_uuid
+        app_name = release.app_name
+        app_dns = release.app_dns
+        version = release.version
+        stories = release.stories
+        maintenance = release.maintenance
+        deleted = release.deleted
+        owner_uuid = release.owner_uuid
+        owner_email = release.owner_email
```
Comment (wilzbach) [inline asyncy/Apps.py]: Hmm I understand the idea of keeping the diff small, but I would have just inserted the `release.` whenever required instead of those 8 lines here.

File: asyncy/Kubernetes.py [hunk 12]
```diff
@@ -560,6 +560,13 @@ def get_liveness_probe(cls, app, service: str):
 
         b16_service_name = base64.b16encode(service_name.encode()).decode()
 
+        if app.always_pull_images is True:
+            image_pull_policy = 'Always'
+        else:
+            image_pull_policy = 'IfNotPresent'
```
Comment (wilzbach) [inline asyncy/Kubernetes.py]: Wouldn't this be better situated as method in `App`?

File: tests/unit/Kubernetes.py [hunk 13]
```diff
@@ -643,6 +645,11 @@ def test_new_ssl_context():
     patch.object(Kubernetes, 'create_imagepullsecret', new=async_mock())
     patch.object(Kubernetes, 'get_liveness_probe', return_value=liveness_probe)
 
+    if story.app.always_pull_images is True:
+        image_pull_policy = 'Always'
+    else:
+        image_pull_policy = 'IfNotPresent'
```
Comment (wilzbach) [inline tests/unit/Kubernetes.py]: If it's `image_pull_policy` is a method you can test it separately and patch it out for this test.
